### PR TITLE
Add fixed partitions

### PR DESCRIPTION
**Pull request by worker24h**
1. Add fixed partitions.  for example:
   ALTER TABLE test_table ADD PARTITION p0125 VALUES [("20190125"), ("20190126"));

ISSUE: #1990

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/catalog/Replica.java [hunk 1]
```diff
@@ -111,7 +111,7 @@ public Replica(long replicaId, long backendId, int schemaHash, ReplicaState stat
     
     // for create tablet and restore
     public Replica(long replicaId, long backendId, ReplicaState state, long version, long versionHash, int schemaHash) {
-        this(replicaId, backendId, version, versionHash, schemaHash, -1, -1, state, -1L, 0L, version, versionHash);
+        this(replicaId, backendId, version, versionHash, schemaHash, 0L, -1, state, -1L, 0L, version, versionHash);
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/catalog/Replica.java]:
```suggestion
        this(replicaId, backendId, version, versionHash, schemaHash, 0L, 0L, state, -1L, 0L, version, versionHash);
```

File: fe/src/main/cup/sql_parser.cup [hunk 2]
```diff
@@ -1515,6 +1521,15 @@ partition_key_desc ::=
     :}
     ;
 
+fixed_partition_key_desc ::=
+    LBRACKET string_list:keys RPAREN
+    {:
+        if (keys.size() != 2) {
```
Comment (morningman) [inline fe/src/main/cup/sql_parser.cup]: I think the proper format is:
`[("low1", "low2"), ("up1", "up2"))`

And should also support :
`[("low1", "low2", ..), ("up1"))`, and the default value should be filled with min value

Do we support MAXVALUE?

**Review by morningman — CHANGES_REQUESTED** [review]
2 more things:
1. Add Documents
2. Unit Test

File: fe/src/main/cup/sql_parser.cup [hunk 3]
```diff
@@ -1515,6 +1523,81 @@ partition_key_desc ::=
...
 
+fixed_partition_key_desc ::=
+    LBRACKET partition_values:keys RPAREN
+    {:
+        if (keys.size() != 2) {
+            throw new Exception("partition range must contains two elements.");
+        }
+        if (keys.get(0).toUpperCase().equals("MAXVALUE")) {
+            throw new Exception("the left value can't MAXVALUE.");
+        }
```
Comment (morningman) [inline fe/src/main/cup/sql_parser.cup]: You should not check semantic here, this should be done in analyze phase in FE.
Here should only check syntax.

File: fe/src/main/cup/sql_parser.cup [hunk 4]
```diff
@@ -1515,6 +1523,81 @@ partition_key_desc ::=
...
+        RESULT = new PartitionKeyDesc(Lists.newArrayList(keys.get(0)), Lists.newArrayList(keys.get(1)));
+    :}
+    | LBRACKET multi_partition_keys_list:keys RPAREN
+      {:
+          if (keys.size() != 2) {
```
Comment (morningman) [inline fe/src/main/cup/sql_parser.cup]: `keys.size() != 2` should already be guaranteed by the syntax your define here.

File: fe/src/main/java/org/apache/doris/analysis/LiteralExpr.java [hunk 5]
```diff
@@ -45,6 +45,9 @@ protected LiteralExpr(LiteralExpr other) {
     public static LiteralExpr create(String value, Type type) throws AnalysisException {
         Preconditions.checkArgument(type != Type.INVALID);
         LiteralExpr literalExpr = null;
+        if (value.toUpperCase().equals("MAXVALUE")) {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/analysis/LiteralExpr.java]: Can not do this!
`MAXVALUE` should be a keyword of syntex, NOT a keyword of LiteralExpr.
What if user create a string "MAXVALUE"?

File: fe/src/main/java/org/apache/doris/catalog/PartitionKey.java [hunk 6]
```diff
@@ -120,11 +120,34 @@ public boolean isMinValue() {
 
     public boolean isMaxValue() {
         for (LiteralExpr literalExpr : keys) {
-            if (literalExpr != MaxLiteral.MAX_VALUE) {
-                return false;
+            if (literalExpr == MaxLiteral.MAX_VALUE) {
+                return true;
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/catalog/PartitionKey.java]: Only all sub keys are MAX_VALUE that this partition key can be MaxValue.
partition key like `(MAX_VALUE, "100")` is not max value.

File: fe/src/main/java/org/apache/doris/catalog/RangePartitionInfo.java [hunk 7]
```diff
@@ -93,12 +93,60 @@ public void addPartition(long partitionId, Range<PartitionKey> range, DataProper
...
+        Range<PartitionKey> newRange = null;
+        PartitionKeyDesc partKeyDesc = desc.getPartitionKeyDesc();
+        try {
+            PartitionKey lowKey =
+                    PartitionKey.createPartitionKey(partKeyDesc.getLowerValues(), partitionColumns);
+            PartitionKey upperKey =
+                    PartitionKey.createPartitionKey(partKeyDesc.getUpperValues(), partitionColumns);
+
+            // check: [left, right), error if left equal right
+            if (lowKey.compareTo(upperKey) == 0) {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/catalog/RangePartitionInfo.java]: `lowKey.compareTo(upperKey) == 0` ==> `lowKey.compareTo(upperKey) >= 0` ?
If this is just for avoid `Range.closedOpen()`'s invalid check, you should add some comment here.

File: fe/src/main/java/org/apache/doris/catalog/RangePartitionInfo.java [hunk 8]
```diff
@@ -93,12 +93,60 @@ public void addPartition(long partitionId, Range<PartitionKey> range, DataProper
...
+                checkRangeIntersect(newRange, nextRange);
+                //keys only have one.
+                int compare = PartitionKey.compareLiteralExpr(upperKey.getKeys().get(0),
+                        nextRange.lowerEndpoint().getKeys().get(0));
+                if (compare <= 0) {
+                    break;//performance optimization
+                }
+            }
+        } catch (AnalysisException e) {
+            throw new DdlException("Invalid range value format： " + e.getMessage());
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/catalog/RangePartitionInfo.java]:
```suggestion
            throw new DdlException("Invalid range value format: " + e.getMessage());
```

File: fe/src/main/java/org/apache/doris/catalog/RangePartitionInfo.java [hunk 9]
```diff
@@ -93,12 +93,60 @@ public void addPartition(long partitionId, Range<PartitionKey> range, DataProper
...
+            }
+
+            // new Range object
+            newRange = Range.closedOpen(lowKey, upperKey);
+
+            // compare
+            for (Map.Entry<Long, Range<PartitionKey>> entry : entries) {
+                Range<PartitionKey> nextRange = entry.getValue();
+                checkRangeIntersect(newRange, nextRange);
+                //keys only have one.
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/catalog/RangePartitionInfo.java]: Why just check first key?

**Review by morningman — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/analysis/PartitionValue.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+import org.apache.doris.common.AnalysisException;
+
+public class PartitionValue {
+    private String value;
+    private boolean isMaxValue;
+
+    public PartitionValue() {
+        isMaxValue = true;
+        value = null;
+    }
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/analysis/PartitionValue.java]: create a function like PartitionValue.createMax() to make it more explicit.

File: fe/src/main/java/org/apache/doris/catalog/PartitionKey.java (hunk 6 above)
Comment (morningman) [reply fe/src/main/java/org/apache/doris/catalog/PartitionKey.java]: You need to modify this

File: fe/src/main/java/org/apache/doris/catalog/RangePartitionInfo.java (hunk 7 above)
Comment (worker24h) [reply fe/src/main/java/org/apache/doris/catalog/RangePartitionInfo.java]: ok

File: fe/src/main/java/org/apache/doris/catalog/PartitionKey.java (hunk 6 above)
Comment (worker24h) [reply fe/src/main/java/org/apache/doris/catalog/PartitionKey.java]: ok

**Review by morningman — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/analysis/PartitionKeyDesc.java [hunk 11]
```diff
@@ -34,84 +34,121 @@
 public class PartitionKeyDesc implements Writable {
     // public static final PartitionKeyDesc MAX_VALUE = new PartitionKeyDesc();
     // lower values only be used for restore
```
[the remark this answers is not in the document]
Comment (worker24h) [reply fe/src/main/java/org/apache/doris/analysis/PartitionKeyDesc.java]: ok

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**